plugins/field-dependent-dropdown/src/field_dependent_dropdown.ts

```diff
@@ -0,0 +1,274 @@
+      // option mapping, use the corresponding options.
+      if (dependencyData.parentField) {
+        const options = optionMapping[dependencyData.parentField.getValue()];
+        if (options) {
+          return options;
+        }
+      }
+
+      // Fall back on basic default options.
+      return [['', '']];
+    };
+
+    super(menuGenerator, validator, config);
+    this.parentName = parentName;
+    this.optionMapping = optionMapping;
+    this.dependencyData = dependencyData;
+  }
+
+  /**
+   * Constructs a FieldDependentDropdown from a JSON arg object.
+   * @param options A JSON object providing "parentName" and "optionMapping".
+   * @returns The new field instance.
+   */
+  static fromJson(
+      options: FieldDependentDropdownFromJsonConfig): FieldDependentDropdown {
+    return new FieldDependentDropdown(
+        options['parentName'],
+        options['optionMapping'],
+        undefined,
+        options);
+  }
+
+  /**
+   * Attach this field to a block.
+   *
+   * @param block The block containing this field.
+   */
+  setSourceBlock(block: Blockly.Block) {
+    super.setSourceBlock(block);
+
+    const parentField: Blockly.Field<string> | null =
+        block.getField(this.parentName);
+
+    if (!parentField) {
+      throw new Error('Could not find a parent field with the name ' +
+          this.parentName + ' for the dependent dropdown.');
+    }
+
+    this.dependencyData.parentField = parentField;
+
+    const oldValidator = parentField.getValidator();
+
+    // A validator function for the parent field that has the side effect of
+    // updating the options of this child dropdown field based on the new value
+    // of the parent field whenever it changes. The validator function is a good
+    // place to do this because it is called immediately while deserializing
+    // workspaces before the following fields are deserialized, so when the
+    // child value is deserialized the appropriate options will already be
+    // available. If the parent already had a validator function, it will be
+    // composed with this one and the new value returned from it will be the
+    // basis for determining the new available options.
+    parentField.setValidator((newValue) => {
+      if (oldValidator) {
+        const validatedValue = oldValidator(newValue);
+        // If a validator returns undefined, that means no change. Otherwise,
+        // use the returned value as the new value.
+        if (validatedValue !== undefined) {
+          newValue = validatedValue;
+        }
+      }
+      this.updateOptionsBasedOnNewValue(newValue);
+      return newValue;
+    });
+    this.updateOptionsBasedOnNewValue(parentField.getValue());
+  }
+
+  /**
+   * Updates the options of this child dropdown field based on the new value of
+   * the parent field.
+   * @param newValue The newly assigned value.
+   */
+  private updateOptionsBasedOnNewValue(newValue: string | undefined): void {
```

Comment:
Is `newValue` actually allowed to be undefined? I don't think it's a valid value that would ever get here, because the class validator for field rejects `null` and `undefined` values, in which case the local validator defined above is never called and the field's value is not updated.

Reply:
The type of FieldValidator is:
export type FieldValidator<T = any> = (value?: T) => T | null | undefined;
...Wait no it's not anymore, it was updated in https://github.com/google/blockly/commit/0fb64a6772bfa3a57c0736f551d96a99fced98ec
export type FieldValidator<T = any> = (newValue: T) => T | null | undefined;
But the argument was still optional in the latest release 9.2.1 of Blockly on npm, and this plugin is configured to depend on ^9.0.0 anyway.